asdf/fits_embed.py

```diff
@@ -20,15 +20,87 @@
 
 try:
     from astropy.io import fits
+    from astropy.io.fits.file import _File
+    from astropy.io.fits.header import Header, _pad_length
 except ImportError:
     raise ImportError("AsdfInFits requires astropy")
 
 
 ASDF_EXTENSION_NAME = 'ASDF'
-
 FITS_SOURCE_PREFIX = 'fits:'
 
 
+__all__ = ['AsdfInFits']
+
+
+class _AsdfHDU(fits.hdu.base.NonstandardExtHDU):
+    """
+    A non-standard extension HDU for encapsulating an entire ASDF file within a
+    single HDU of a container FITS file.  These HDUs have an extension (that is
+    an XTENSION keyword) of ASDF.
+    """
+
+    _extension = ASDF_EXTENSION_NAME
+
+    @classmethod
+    def from_buff(cls, buff, compress=False, **kwargs):
+        """
+        Creates a new _AsdfHDU from a given AsdfFile object.
+
+        Parameters
+        ----------
+        buff : io.BytesIO
+            A buffer containing an ASDF metadata tree
+        compress : bool, optional
+            Gzip compress the contents of the ASDF HDU
+        """
+
+        if compress:
+            buff = gzip.GzipFile(fileobj=buff, mode='wb')
+
+        # A proper HDU should still be padded out to a multiple of 2880
+        # technically speaking
+        data_length = buff.tell()
+        padding = (_pad_length(data_length) * cls._padding_byte).encode('ascii')
+        buff.write(padding)
+
+        buff.seek(0)
+
+        cards = [
+            ('XTENSION', cls._extension, 'ASDF extension'),
+            ('BITPIX', 8, 'array data type'),
+            ('NAXIS', 1, 'number of array dimensions'),
+            ('NAXIS1', len(buff.getvalue()), 'Axis length'),
+            ('PCOUNT', 0, 'number of parameters'),
+            ('GCOUNT', 1, 'number of groups'),
+            ('COMPRESS', compress, 'Uses gzip compression'),
+            ('EXTNAME', cls._extension, 'Name of ASDF extension'),
```

Comment:
If I read this right, the *name* is still hard-coded as `ASDF`.  Is that desirable?  I would have thought this should allow multiple ASDF-type extensions, with arbitrary names.  Or is this necessary for backwards compatibility?

Reply:
Admittedly this is not the most generic design. However, since the `AsdfInFits` class is expected to be the only user of this HDU class, and since `astropy.io.fits` keys HDUs in an `HDUList` based on `EXTNAME`, it just made sense to set `EXTNAME` to `ASDF` here. In other words, this is a matter of convenience and an implementation detail of `AsdfInFits`, but not necessarily part of the abstract ASDF HDU convention itself.

Reply:
Multiple ASDF extensions could be distinguished via unique EXTVER values, assuming that keyword makes sense in an ASDF extension?

Reply:
It might be possible, but I would need to investigate more. Currently `AsdfInFits` expects only a single ASDF extension to be present in a FITS file. Until the JWST pipeline has a use case for multiple ASDF extensions in the same file, I don't think it makes sense to support this.

Reply:
Ah, well, I'm not necessarily attached to this *specific* line as the spot to change it.  Are you just thinking that if I am a user and want to have an ASDF-in-FITS HDU with the name "WCS_MODEL", I'd be expected to do something like
```
fits_embed.AsdfInFits(hdulist, tree)
hdulist['ASDF'].header['EXTNAME] = 'WCS_MODEL'
```
?  I'm really just saying I think it would be useful to have the option to do something like ``fits_embed.AsdfInFits(hdulist, tree, name='WCS_MODEL')``, as the typical user is likely to see that as the main place where they set information about the hdu.

Reply:
Oh, I think I just realized my mis-conception.  I thought it was possible to have *multiple* asdf trees in one file. I see now that the "reader" mode only supports a single one.  I guess that makes this use case less important.  Probably I should make a *separate* issue about that, but this concern about the naming could wait until then...

Reply:
Just to be clear: the ASDF HDU is used solely for storing a serialized version of the ASDF metadata tree. `AsdfInFits` currently expects a one-to-one mapping between FITS files and ASDF trees. The actual data arrays, including any used for WCS models or any other purposes, are actually stored in conventional FITS HDUs. The details of this are controlled completely by the user.

So it doesn't really make sense to use anything other than `ASDF` as the `EXTNAME` for this HDU because the only thing that will ever actually process it is our `AsdfInFits` code. Its just an implementation detail and anyone who reads the pure FITS file does not need to be concerned with it. Unless there becomes a compelling reason to have more than one ASDF metadata tree in the same FITS file, I don't think it makes sense to change anything.

Reply:
Gotcha, that clears things up.  I withdraw my request 😉